src/index.ts

```diff
@@ -1,31 +1,181 @@
-const REGION_IDS = ["auk", "bop", "can", "cit", "gis", "hkb", "mwt", "mbh", "nsn", "ntl", "ota", "stl", "tki", "tas", "wko", "wgn", "wtc"]
+const REGION_IDS = ["AUK", "BOP", "CAN", "CIT", "GIS", "HKB", "MBH", "MWT", "NSN", "NTL", "OTA", "STL", "TKI", "TAS", "WKO", "WGN", "WTC"];
 
 interface DistrictData {
-  color: string
-  description?: string
+  color?: string,
+  descriptions?: string
 }
 
 interface MapOptions {
   width?: number,
   height?: number,
+  targetSelector: string,
+  descriptions?: {
+    [key: string]: string;
+  },
+  colors?: {
+    [key: string]: string;
+  },
   data?: {
     [key in typeof REGION_IDS[number]]: DistrictData
   }
 }
 
 class NZMapSVG {
-  options: MapOptions
+  options: MapOptions;
+
+  // Element tags
+  wrapper: any; //<div>
+  mapImage: any; //<svg>
+  descriptionDisplay: any //<p>
 
   constructor(options: MapOptions) {
-    this.options = options
+    this.options = options;
   }
 
   // Add other methods here
+  render() {
+    if (
+      !this.options.targetSelector ||
+      !document.querySelector(`#${this.options.targetSelector}`)
+    ) {
+      this.error('Target element not found');
+    }
+
+    // <svg> inside <div id='map'>
+    this.wrapper = document.querySelector(`#${this.options.targetSelector}`);
+    this.mapImage = document.createElementNS(
+      'http://www.w3.org/2000/svg',
+      'svg'
+    );
+    this.wrapper.appendChild(this.mapImage);
+
+    this.mapImage.setAttribute('viewBox', '0 0 701.42358 988.42462');
+    this.mapImage.setAttribute('fill', 'white');
+    this.mapImage.setAttribute('stroke', 'black');
+    this.mapImage.classList.add('svgMap-map-image');
+
+    // descriptionDisplay p tag
+    this.descriptionDisplay = document.createElement('p');
+    this.descriptionDisplay.setAttribute('style', 'font-size: 80px');
+    this.wrapper.appendChild(this.descriptionDisplay);
+
+
+    // <path> inside <svg>
+    for (let regionID in this.mapPaths) {
+      var regionData = this.mapPaths[regionID];
+      if (!regionData.d) {
+        return;
+      }
+
+      var regionElement = document.createElementNS(
+        'http://www.w3.org/2000/svg',
+        'path'
+      );
+
+      regionElement.setAttribute('d', regionData.d);
+      regionElement.setAttribute(
+        'id',
+        regionID
+      );
+      regionElement.setAttribute('data-id', regionID);
+      regionElement.classList.add('svgMap-country');
+
+      // render in specific colours:
+      regionElement.setAttribute('style', {'color': this.options.colors[regionID]});
+      this.mapImage.appendChild(regionElement);
+
+      // Test for hover code:
+      regionElement.addEventListener(
+        'mouseover',
+        (e) => {
+          let regionID = e.target.id;
+          let regionDescription = this.options.descriptions[regionID];
+          console.log(regionID +': '+ regionDescription);
+          this.descriptionDisplay.innerHTML = regionID +': '+ regionDescription
```

Comment:
We should be weary whenever we try set a string to HTML through innerHTML, we can use textContent instead, see: https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent

The main difference is textContent just renders the string as text which prevents XSS attacks.
E.g. a malicious chrome extension might set the region id tag to `<iframe src="mybank.com/transfer?to=kewei&amount=10000"/>`, which could have bad consequences if the browser rendered that as HTML.

Reply:
ok changed